File: native/android/VueNativeCore/src/main/kotlin/com/vuenative/core/Bridge/JSRuntime.kt
package com.vuenative.core

import android.content.Context
import android.os.Handler
import android.os.HandlerThread
import android.os.Looper
import android.util.Log
import com.eclipsesource.v8.JavaVoidCallback
import com.eclipsesource.v8.V8

/**
 * Core JavaScript runtime. Wraps J2V8's V8 engine on a dedicated HandlerThread.
 *
 * Thread safety contract:
 * - All V8 access happens exclusively on jsThread via jsHandler
 * - Never pass J2V8 objects (V8Array, V8Object) across threads
 * - All View operations happen on main thread via mainHandler
 */
class JSRuntime(private val context: Context) {

    companion object {
        private const val TAG = "VueNative-JSRuntime"
    }

    /** Dedicated JS thread. All V8 operations run here. */
    private val jsThread = HandlerThread("VueNative-JS").apply { start() }

    /** Handler for posting work to the JS thread. */
    val jsHandler = Handler(jsThread.looper)

    /** Handler for posting work to the main (UI) thread. */
    private val mainHandler = Handler(Looper.getMainLooper())

    /** V8 runtime instance. Only access from jsThread. */
    private var v8: V8? = null

    /** Whether the runtime has been initialized. */
    private var isInitialized = false

    /** The NativeBridge instance — manages view registry and operations. */
    val bridge = NativeBridge(context)

    /** Startup time in milliseconds for performance.now(). */
    val startTimeMs: Long = System.currentTimeMillis()

    init {
        // Wire bridge callbacks
        bridge.onFireEvent = { nodeId, eventName, payloadJson ->
            // Called on main thread — dispatch to JS
            jsHandler.post {
                try {
                    if (nodeId == -1 && eventName == "__callback__") {
                        // Native module async callback — route to __VN_resolveCallback
                        v8?.executeVoidScript(
                            "if(typeof __VN_resolveCallback==='function'){" +
                            "var _d=$payloadJson;" +
                            "__VN_resolveCallback(_d.callbackId,_d.result,_d.error);}"
                        )
                    } else {
                        v8?.executeVoidScript(
                            "if(typeof __VN_handleEvent==='function')" +
                            "__VN_handleEvent($nodeId,${encodeJs(eventName)},$payloadJson)"
                        )
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "Error dispatching event $eventName to JS", e)
                }
            }
        }
        bridge.onDispatchGlobalEvent = { eventName, payloadJson ->
            jsHandler.post {
                try {
                    v8?.executeVoidScript(
                        "if(typeof __VN_handleGlobalEvent==='function')" +
                        "__VN_handleGlobalEvent(${encodeJs(eventName)},${encodeJs(payloadJson)})"
                    )
                } catch (e: Exception) {
                    Log.e(TAG, "Error dispatching global event $eventName", e)
                }
            }
        }
    }

    /** Initialize the V8 runtime on the JS thread. Calls completion when ready. */
    fun initialize(completion: (() -> Unit)? = null) {
        jsHandler.post {
            if (isInitialized) {
                mainHandler.post { completion?.invoke() }
                return@post
            }
            try {
                v8 = V8.createV8Runtime()
                JSPolyfills.register(this)
                // Register __VN_flushOperations — JS calls this to send batched ops to native
                v8?.registerJavaMethod(JavaVoidCallback { _, params ->
                    try {
                        val json = if (params.length() > 0) params.getString(0) else "[]"
                        bridge.processOperations(json)
                    } finally {
                        params.close()
                    }
                }, "__VN_flushOperations")

                // Register __VN_handleError — JS calls this to report errors to native
                v8?.registerJavaMethod(JavaVoidCallback { _, params ->
                    try {
                        val errorJson = if (params.length() > 0) params.getString(0) else "{}"
                        Log.e(TAG, "[VueNative Error] $errorJson")
                    } finally {
                        params.close()
                    }
                }, "__VN_handleError")

                isInitialized = true
                Log.d(TAG, "V8 runtime initialized")
                mainHandler.post { completion?.invoke() }
            } catch (e: Exception) {
                Log.e(TAG, "Failed to initialize V8 runtime", e)
            }
        }
    }

    /** Load and execute a JavaScript bundle. Must be called after initialize().
     *  @param onComplete Optional callback invoked on main thread: (success, errorMessage). */
    fun loadBundle(bundleCode: String, onComplete: ((Boolean, String?) -> Unit)? = null) {
        jsHandler.post {
            try {
                v8?.executeVoidScript(bundleCode)
                Log.d(TAG, "Bundle loaded successfully")
                mainHandler.post { onComplete?.invoke(true, null) }
            } catch (e: Exception) {
                Log.e(TAG, "Error loading bundle", e)
                val msg = e.message ?: "Unknown error"
                mainHandler.post {
                    ErrorOverlayView.show(context, msg)
                    onComplete?.invoke(false, msg)
                }
            }
        }
    }

    /** Post work to the JS thread. */
    fun runOnJsThread(block: () -> Unit) {
        jsHandler.post(block)
    }

    /** Post work to the main (UI) thread. */
    fun runOnMainThread(block: () -> Unit) {
        mainHandler.post(block)
    }

    /** Execute a void script on the JS thread. Safe to call from any thread. */
    fun executeVoidScript(script: String) {
        jsHandler.post {
            try {
                v8?.executeVoidScript(script)
            } catch (e: Exception) {
                Log.e(TAG, "Script error: ${e.message}")
            }
        }
    }

    /** Register a void Java method as a global JS function. Call from JS thread only. */
    fun registerVoidCallback(name: String, callback: (Array<Any?>) -> Unit) {
        jsHandler.post {
            v8?.registerJavaMethod(JavaVoidCallback { _, params ->
                try {
                    val args = Array<Any?>(params.length()) { i ->
                        when {
                            params.getType(i) == 1 -> params.getInteger(i) // INT
                            params.getType(i) == 2 -> params.getDouble(i) // DOUBLE
                            params.getType(i) == 3 -> params.getBoolean(i) // BOOLEAN
                            params.getType(i) == 4 -> params.getString(i) // STRING
                            else -> null
                        }
                    }
                    callback(args)
                } finally {
                    params.close()
                }
            }, name)
        }
    }

    /** Access raw V8 instance — only call from JS thread! */
    fun withV8(block: (V8) -> Unit) {
        jsHandler.post {
            v8?.let(block)
        }
    }

    /** Get raw V8 — ONLY call from JS thread. Returns null if not initialized. */
    fun v8(): V8? = v8

    /** Release the V8 runtime. */
    fun release() {
        jsHandler.post {
            v8?.release(true)
            v8 = null
            isInitialized = false
        }
        jsThread.quitSafely()
    }

    private fun encodeJs(value: String): String =
        "\"${value.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n").replace("\r", "\\r")}\""
}

File: native/shared/VueNativeShared/Sources/VueNativeShared/JSRuntime.swift
import JavaScriptCore
import Foundation

// MARK: - Bundle Source

/// Describes where to load the JS application bundle from.
public enum BundleSource {
    /// Load from an embedded resource in the app bundle.
    case embedded(name: String)
    /// Load from a development server URL (for live reload).
    case devServer(url: URL)
}

// MARK: - JSRuntime

/// Core JavaScript runtime manager. Wraps JSContext on a dedicated serial DispatchQueue.
/// All JS operations are guaranteed to execute on the JS queue. UI operations
/// are never performed on this queue.
///
/// Thread safety contract:
/// - All JSContext access happens exclusively on `jsQueue`
/// - Never pass JSValue across threads — extract primitives first
/// - All closures registered with JSContext use [weak self]
public final class JSRuntime: @unchecked Sendable {

    // MARK: - Singleton

    public static let shared = JSRuntime()

    // MARK: - Properties

    /// Dedicated serial queue for all JavaScript execution.
    /// QoS is userInteractive because JS drives the UI pipeline.
    public let jsQueue = DispatchQueue(label: "com.vuenative.js", qos: .userInteractive)

    /// The underlying JavaScriptCore context. Only access on jsQueue.
    public private(set) var context: JSContext!

    /// Whether the runtime has been initialized.
    public private(set) var isInitialized = false

    /// Startup time reference for performance.now()
    public let startTime: CFAbsoluteTime = CFAbsoluteTimeGetCurrent()

    /// Callback invoked in DEBUG builds when JS throws an exception.
    /// Platform-specific layers set this to show an error overlay.
    public var onError: ((String) -> Void)?

    // MARK: - Initialization

    private init() {}

    /// Initialize the JS runtime. Creates the JSContext on the JS queue,
    /// configures exception handling, and registers polyfills.
    /// Must be called before any other method.
    public func initialize(completion: (() -> Void)? = nil) {
        jsQueue.async { [weak self] in
            guard let self = self else { return }
            guard !self.isInitialized else {
                completion?()
                return
            }

            // Create the JSContext
            self.context = JSContext()

            // Configure exception handler
            self.context.exceptionHandler = { [weak self] context, exception in
                guard let exception = exception else { return }
                let message = exception.toString() ?? "Unknown JS error"
                let line = exception.objectForKeyedSubscript("line")?.toInt32() ?? 0
                let column = exception.objectForKeyedSubscript("column")?.toInt32() ?? 0
                let stack = exception.objectForKeyedSubscript("stack")?.toString() ?? ""
                NSLog("[VueNative JS Error] \(message) at line \(line):\(column)")
                if !stack.isEmpty {
                    NSLog("[VueNative JS Stack] \(stack)")
                }
                #if DEBUG
                let fullMessage = stack.isEmpty ? message : "\(message)\n\n\(stack)"
                self?.onError?(fullMessage)
                #endif
                _ = self // prevent unused warning
            }

            // Set globalThis = global object
            self.context.evaluateScript("var globalThis = this;")

            // Register shared polyfills (everything except platform-specific RAF)
            SharedJSPolyfills.register(in: self)

            self.isInitialized = true
            completion?()
        }
    }

    // MARK: - Script Evaluation

    /// Evaluate a JavaScript string on the JS queue.
    /// The completion handler is called on the JS queue with the result.
    public func evaluateScript(_ script: String, sourceURL: String? = nil, completion: ((JSValue?) -> Void)? = nil) {
        jsQueue.async { [weak self] in
            guard let self = self, let context = self.context else {
                completion?(nil)
                return
            }

            let result: JSValue?
            if let sourceURL = sourceURL, let url = URL(string: sourceURL) {
                result = context.evaluateScript(script, withSourceURL: url)
            } else {
                result = context.evaluateScript(script)
            }

            // Force microtask drain after evaluation.
            // JSC may not automatically drain microtasks after evaluateScript returns.
            // This ensures Vue's Promise-based scheduler flushes.
            context.evaluateScript("void 0;")

            completion?(result)
        }
    }

    /// Evaluate a script synchronously. MUST only be called from the JS queue.
    /// Returns the JSValue result directly.
    @discardableResult
    public func evaluateScriptSync(_ script: String, sourceURL: String? = nil) -> JSValue? {
        dispatchPrecondition(condition: .onQueue(jsQueue))
        guard let context = context else { return nil }

        let result: JSValue?
        if let sourceURL = sourceURL, let url = URL(string: sourceURL) {
            result = context.evaluateScript(script, withSourceURL: url)
        } else {
            result = context.evaluateScript(script)
        }

        // Force microtask drain
        context.evaluateScript("void 0;")
        return result
    }

    // MARK: - Function Calls

    /// Call a global JS function by name with arguments. Runs on the JS queue.
    /// Uses objectForKeyedSubscript().call() for performance (avoids re-parsing).
    public func callFunction(_ name: String, withArguments args: [Any], completion: ((JSValue?) -> Void)? = nil) {
        jsQueue.async { [weak self] in
            guard let self = self, let context = self.context else {
                completion?(nil)
                return
            }

            let function = context.objectForKeyedSubscript(name)
            guard let fn = function, !fn.isUndefined else {
                NSLog("[VueNative] Warning: JS function '\(name)' not found")
                completion?(nil)
                return
            }

            let result = fn.call(withArguments: args)

            // Force microtask drain
            context.evaluateScript("void 0;")

            completion?(result)
        }
    }

    /// Call a global JS function synchronously. MUST only be called from the JS queue.
    @discardableResult
    public func callFunctionSync(_ name: String, withArguments args: [Any]) -> JSValue? {
        dispatchPrecondition(condition: .onQueue(jsQueue))
        guard let context = context else { return nil }

        let function = context.objectForKeyedSubscript(name)
        guard let fn = function, !fn.isUndefined else {
            NSLog("[VueNative] Warning: JS function '\(name)' not found")
            return nil
        }

        let result = fn.call(withArguments: args)

        // Force microtask drain
        context.evaluateScript("void 0;")
        return result
    }

    // MARK: - Function Registration

    /// Register a Swift function as a global JS function.
    /// The block receives a single JSValue argument (use .toArray() etc. to extract args).
    /// IMPORTANT: The block executes on the JS queue.
    public func registerFunction(_ name: String, block: @escaping @convention(block) (JSValue) -> Void) {
        jsQueue.async { [weak self] in
            guard let self = self, let context = self.context else { return }
            let wrappedBlock: @convention(block) (JSValue) -> Void = { [weak self] value in
                _ = self // prevent retain cycle through closure capture
                block(value)
            }
            context.setObject(wrappedBlock, forKeyedSubscript: name as NSString)
        }
    }

    /// Register a Swift function with multiple arguments.
    public func registerFunctionMultiArg(_ name: String, argCount: Int, block: @escaping ([JSValue]) -> Void) {
        jsQueue.async { [weak self] in
            guard let self = self, let context = self.context else { return }

            switch argCount {
            case 0:
                let wrapped: @convention(block) () -> Void = { [weak self] in
                    _ = self
                    block([])
                }
                context.setObject(wrapped, forKeyedSubscript: name as NSString)
            case 1:
                let wrapped: @convention(block) (JSValue) -> Void = { [weak self] a in
                    _ = self
                    block([a])
                }
                context.setObject(wrapped, forKeyedSubscript: name as NSString)
            case 2:
                let wrapped: @convention(block) (JSValue, JSValue) -> Void = { [weak self] a, b in
                    _ = self
                    block([a, b])
                }
                context.setObject(wrapped, forKeyedSubscript: name as NSString)
            case 3:
                let wrapped: @convention(block) (JSValue, JSValue, JSValue) -> Void = { [weak self] a, b, c in
                    _ = self
                    block([a, b, c])
                }
                context.setObject(wrapped, forKeyedSubscript: name as NSString)
            default:
                // For 4+ args, use the single-arg version and extract from JSValue
                let wrapped: @convention(block) (JSValue) -> Void = { [weak self] value in
                    _ = self
                    block([value])
                }
                context.setObject(wrapped, forKeyedSubscript: name as NSString)
            }
        }
    }

    // MARK: - Bundle Loading

    /// Load a JS application bundle from the given source.
    /// Polyfills are already registered during initialize().
    public func loadBundle(source: BundleSource, completion: ((Bool) -> Void)? = nil) {
        jsQueue.async { [weak self] in
            guard let self = self, self.context != nil else {
                completion?(false)
                return
            }

            switch source {
            case .embedded(let name):
                self.loadEmbeddedBundle(name: name, completion: completion)

            case .devServer(let url):
                self.loadDevServerBundle(url: url, completion: completion)
            }
        }
    }

    // MARK: - Private: Bundle Loading

    private func loadEmbeddedBundle(name: String, completion: ((Bool) -> Void)?) {
        // Try to find the bundle in the main app bundle
        let bundleName: String
        let bundleExtension: String

        if name.contains(".") {
            let parts = name.split(separator: ".", maxSplits: 1)
            bundleName = String(parts[0])
            bundleExtension = String(parts[1])
        } else {
            bundleName = name
            bundleExtension = "js"
        }

        // Search in main bundle
        let scriptURL = Bundle.main.url(forResource: bundleName, withExtension: bundleExtension)

        guard let url = scriptURL else {
            NSLog("[VueNative] Error: Bundle '\(name)' not found in app resources")
            completion?(false)
            return
        }

        do {
            let script = try String(contentsOf: url, encoding: .utf8)
            NSLog("[VueNative] Loading bundle: \(name) (\(script.count) bytes)")
            self.context.evaluateScript(script, withSourceURL: url)
            // Force microtask drain after bundle load
            self.context.evaluateScript("void 0;")
            NSLog("[VueNative] Bundle loaded successfully")
            completion?(true)
        } catch {
            NSLog("[VueNative] Error loading bundle '\(name)': \(error.localizedDescription)")
            completion?(false)
        }
    }

    private func loadDevServerBundle(url: URL, completion: ((Bool) -> Void)?) {
        // Fetch from dev server on a background queue, then evaluate on JS queue
        let task = URLSession.shared.dataTask(with: url) { [weak self] data, response, error in
            guard let self = self else {
                completion?(false)
                return
            }

            self.jsQueue.async { [weak self] in
                guard let self = self, self.context != nil else {
                    completion?(false)
                    return
                }

                if let error = error {
                    NSLog("[VueNative] Dev server error: \(error.localizedDescription)")
                    completion?(false)
                    return
                }

                guard let data = data, let script = String(data: data, encoding: .utf8) else {
                    NSLog("[VueNative] Dev server returned invalid data")
                    completion?(false)
                    return
                }

                NSLog("[VueNative] Loading dev bundle from \(url) (\(script.count) bytes)")
                self.context.evaluateScript(script, withSourceURL: url)
                // Force microtask drain
                self.context.evaluateScript("void 0;")
                NSLog("[VueNative] Dev bundle loaded successfully")
                completion?(true)
            }
        }
        task.resume()
    }

    // MARK: - Hot Reload

    /// Reload the runtime with a new JavaScript bundle string.
    /// Creates a fresh JSContext, re-registers polyfills, and evaluates the bundle.
    /// Calls completion on the JS queue with success/failure.
    public func reload(bundle: String, completion: ((Bool) -> Void)? = nil) {
        jsQueue.async { [weak self] in
            guard let self = self else {
                completion?(false)
                return
            }

            NSLog("[VueNative] Hot reload: tearing down old context...")

            // Tear down old Vue app gracefully
            if let teardown = self.context?.objectForKeyedSubscript("__VN_teardown"),
               !teardown.isUndefined {
                teardown.call(withArguments: [])
                self.context?.evaluateScript("void 0;")
            }

            // Reset polyfill state (timers) before creating new context
            SharedJSPolyfills.reset()

            // Create fresh context
            self.context = JSContext()
            self.context.exceptionHandler = { [weak self] _, exception in
                guard let exception = exception else { return }
                let message = exception.toString() ?? "Unknown JS error"
                let line = exception.objectForKeyedSubscript("line")?.toInt32() ?? 0
                let stack = exception.objectForKeyedSubscript("stack")?.toString() ?? ""
                NSLog("[VueNative JS Error] \(message) at line \(line)")
                #if DEBUG
                let fullMessage = stack.isEmpty ? message : "\(message)\n\n\(stack)"
                self?.onError?(fullMessage)
                #endif
                _ = self
            }

            // Re-register globalThis and polyfills
            self.context.evaluateScript("var globalThis = this;")
            SharedJSPolyfills.register(in: self)

            NSLog("[VueNative] Hot reload: evaluating new bundle (\(bundle.count) bytes)...")
            self.context.evaluateScript(bundle)
            if let exception = self.context.exception {
                NSLog("[VueNative] Hot reload bundle error: %@", exception.toString() ?? "unknown")
                self.context.exception = nil
                completion?(false)
                return
            }
            self.context.evaluateScript("void 0;")
            NSLog("[VueNative] Hot reload: complete")

            completion?(true)
        }
    }

    // MARK: - Teardown

    /// Invalidate the runtime and release resources.
    /// After calling this, the runtime cannot be used again.
    public func invalidate() {
        jsQueue.async { [weak self] in
            guard let self = self else { return }
            self.context = nil
            self.isInitialized = false
            NSLog("[VueNative] JS runtime invalidated")
        }
    }
}
